draft-gpew-priv-ppm.md

```diff
@@ -672,247 +677,375 @@ independently.
+1. For each valid report share, initialize the VDAF preparation process.
+
+An invalid report share is marked with one of the following errors:
+
+~~~
+enum {
+  batch-collected(0),
+  report-replayed(1),
+  report-dropped(2),
+  hpke-unknown-config-id(3),
+  hpke-decrypt-error(4),
+  vdaf-prep-error(5),
+} ReportShareError;
+~~~
+
+#### Leader Initialization
+
+The leader begins the aggregate initialization phase with the set of candidate report
+shares as follows:
+
+1. Decrypt the input share for each report share as described in {{input-share-decryption}}.
+1. Check that the resulting input share is valid as described in {{input-share-batch-validation}}.
+1. Initialize VDAF preparation as described in {{input-share-prep}}.
+
+If any step yields an invalid report share, the leader removes the report share from
+the set of candidate reports. Once the leader has initialized this state for all valid
+candidate report shares, it then creates an AggregateInitReq message for each helper to
+initialize the preparation of this candidate set. This message is structured as follows:
 
 ~~~
+struct {
+  Nonce nonce;
+  Extension extensions<0..2^16-1>;
+  HpkeCiphertext encrypted_input_share;
+} ReportShare;
+
 struct {
   TaskID task_id;
-  opaque agg_param<0..2^16-1>;  // VDAF aggregation parameter
-  opaque helper_state<0..2^16>; // helper's opaque state
-  AggregateSubReq seq<1..2^24-1>;
-} AggregateReq;
+  opaque agg_param<0..2^16-1>;
+  opaque helper_state<0..2^16>;
+  ReportShare report_shares<1..2^16-1>;
+} AggregateInitReq;
 ~~~
 
-The structure contains the PPM task, an opaque, VDAF-specific aggregation
-parameter, an opaque *helper state* string, and a sequence of *sub-requests*,
-each corresponding to a unique client report. Sub-requests are structured as
-follows:
+The `nonce` and `extensions` fields of each ReportShare match that in the Report
+uploaded by the client. The `encrypted_input_share` field is the `HpkeCiphertext`
+whose index in `Report.encrypted_input_shares` is equal to the index of the aggregator
+in the task's `aggregator_endpoints`. The `agg_param` field is an opaque, VDAF-specific
+aggregation parameter. The `helper_state` parameter contains the helper's state.
+This is an optional parameter of an aggregate request that the helper can use to carry
+state across requests and across aggregate flows.
+
+Let `[aggregator]` denote the helper's API endpoint. The leader sends a POST
+request to `[aggregator]/aggregate` with its AggregateInitReq message as
+the payload. The media type is "message/ppm-aggregate-init-req".
+
+#### Helper Initialization
+
+Each helper begins their portion of the aggregate initialization phase with the set
+of candidate report shares obtained in an `AggregateInitReq` message from the leader.
+It attempts to recover and validate the corresponding input shares similar to the leader,
+and eventually returns a response to the leader carrying a VDAF-specific message for each
+report share.
```

Comment:
I wonder if it would be better to factor out the common behaviors for all aggregators including the leader.

Reply:
I think this done right now. The commonality is these steps, each of which have their own sections:

```
1. Decrypt the input share for each report share as described in {{input-share-decryption}}.
1. Check that the resulting input share is valid as described in {{input-share-batch-validation}}.
1. Initialize VDAF preparation and initial outputs as described in {{input-share-prep}}.
```